### ISSUE TITLE
Casts on handles

### ISSUE THREAD

**Issue by ktbarrett** [cocotb/cocotb#1222]
Most leaf handle types (`ConstantObject` and `NonConstantObject` and it's children) have cast overloads associated with them which do the convenience behavior of `cast(handle.value)`. This works for `int`, `str`, and `float` on objects that support them. What's missing is `bool`, it was assumed because the previous worked, that a `bool` cast on a VHDL `Boolean` object would as well. Instead it uses the default behavior for bool on a value which means it always evaluates to `True`.

I don't think we should add `__bool__`, but instead remove the other interfaces to maintain consistency. The actual operation it offers is trivial. Forcing the users to use `handle.value` instead makes reading and writing to handles more obvious when reviewing source code.

**Comment by eric-wieser** [comment]
To be clear, you're suggesting asking the user to change `int(dut.signal)` to `int(dut.signal.value)`?

We definitely shouldn't force users to change, but we could start emitting warnings if they don't.

I suspect this might be really quite intrusive though.

**Comment by ktbarrett** [comment]
> To be clear, you're suggesting asking the user to change int(dut.signal) to int(dut.signal.value)?

Yes, and that would be the extent of the change. For `int`, `str` and `float`. 

> I suspect this might be really quite intrusive though.

It likely would be. The alternative is we could add `__bool__`. Like the `int` cast it would fail if the signal was non-0/1 and return `True` if the signal was all '1' and '0' otherwise, just like C's bool cast behavior. Of course, then anyone using the following idiom would silently start doing the wrong thing.

```python
handle = dut.signal if check() else None
# ...
if handle:
    stuff()
```

If we do nothing, this code will silently do the wrong thing.

```python
if bool(dut.fifo.ready):
    dut.fifo.data <= data
    dut.fifo.en <= 1
```

This will never silently do the wrong thing.

```python
if bool(dut.fifo.ready.value):
    dut.fifo.data.value = data
    dut.fifo.en.value = 1

**Comment by eric-wieser** [comment]
Note that defining `__bool__` to raise or warn is a very easy way to stop either party from doing the wrong thing silently

**Comment by ktbarrett** [comment]
With that solution, cocotb would complain always in the first case, even though that usage is technically correct. The warning could mention "use `if handle is not None:`", but someone has to change their code to not get an earful.

**Comment by ktbarrett** [comment]
@eric-wieser I ran into another bug today because of the behavior of the bool cast, I'd like to find a solution. Would you be fine with a definition of `__bool__` being `bool(int(self))` and also raise a warning that the semantics of `__bool__` have changed?

**Comment by eric-wieser** [comment]
Typically warnings are used to indicate that something will change soon, not that it has already changed.

> but with any solution someone has to change their code to not get an earful.

Correct, and that's kind of the point - that earful is what motivates users to change their code before it becomes wrong.

I'm tempted to add
```python
def __bool__(self):
    warnings.warn(
        "`bool(handle)` is deprecated. To keep the old behavior of checking if the handle is `None`, use `handle is not None`. If your intent was to check the value of the signal, use `bool(handle.value)` instead. In future, this will be an error.", DeprecationWarning, stacklevel=2)
```

**Comment by ktbarrett** [comment]
I'm not a fan of the inconsistency, so I tried several things out. Turns out even internal code `handle.__hasattr__` depends upon the current behavior.

**Comment by ktbarrett** [comment]
#1510 fixes the issue from the last comment. Willrevisit this.
